### PR TITLE
Cancel Activity Sample

### PR DESCRIPTION
**Pull request by longquanzheng**
longer@~/gocode/src/github.com/samarabbas/cadence-samples:(cancel)$ ./bin/cancelactivity -m worker
2019-07-01T11:33:14.344-0700	INFO	common/sample_helper.go:61	Logger created.
2019-07-01T11:33:14.345-0700	DEBUG	common/factory.go:116	Creating RPC dispatcher outbound	{"ServiceName": "cadence-frontend", "HostPort": "127.0.0.1:7933"}
2019-07-01T11:33:14.352-0700	INFO	common/sample_helper.go:79	Domain succeesfully registered.	{"Domain": "samples-domain"}
2019-07-01T11:33:14.391-0700	INFO	internal/internal_worker.go:1036	Started Worker	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:14.401-0700	INFO	cancelactivity/workflow.go:39	cancel workflow started	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b"}
2019-07-01T11:33:14.401-0700	DEBUG	internal/internal_event_handlers.go:400	ExecuteActivity	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b", "ActivityID": "0", "ActivityType": "main.activityToBeCacneled"}
2019-07-01T11:33:14.410-0700	INFO	cancelactivity/workflow.go:62	activity started, you can use ./cancelactivity -m cancel <WorkflowID> or CLI: 'cadence --do samples-domain wf cancel -w <WorkflowID>' to cancel	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:15.410-0700	INFO	cancelactivity/workflow.go:66	heartbeating...	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:16.416-0700	INFO	cancelactivity/workflow.go:66	heartbeating...	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:17.419-0700	INFO	cancelactivity/workflow.go:66	heartbeating...	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:17.664-0700	DEBUG	internal/internal_event_handlers.go:418	RequestCancelActivity	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b", "ActivityID": "0"}
2019-07-01T11:33:18.420-0700	INFO	cancelactivity/workflow.go:66	heartbeating...	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:19.420-0700	INFO	cancelactivity/workflow.go:69	context is cancelled	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}
2019-07-01T11:33:19.428-0700	INFO	cancelactivity/workflow.go:51	activityToBeCacneled returns I am canceled by Done, <nil>{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b"}
2019-07-01T11:33:19.428-0700	DEBUG	internal/internal_event_handlers.go:400	ExecuteActivity	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b", "ActivityID": "1", "ActivityType": "main.activityToBeSkipped"}
2019-07-01T11:33:19.428-0700	DEBUG	internal/internal_event_handlers.go:418	RequestCancelActivity	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b", "ActivityID": "1"}
2019-07-01T11:33:19.428-0700	INFO	cancelactivity/workflow.go:55	Workflow completed.	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b"}
2019-07-01T11:33:19.428-0700	DEBUG	internal/internal_event_handlers.go:400	ExecuteActivity	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup", "WorkflowType": "main.Workflow", "WorkflowID": "cancel_171e59b6-d757-491e-b8dd-6b22d9e3bd1e", "RunID": "1247e4a0-78a9-4727-b84a-9a2f14ed4f3b", "ActivityID": "2", "ActivityType": "main.cleanupActivity"}
2019-07-01T11:33:19.435-0700	INFO	cancelactivity/workflow.go:77	cleanupActivity started	{"Domain": "samples-domain", "TaskList": "cancelGroup", "WorkerID": "11687@longer-C02V60N3HTDG@cancelGroup"}

### CODE REVIEW

File: cmd/samples/recipes/cancel/cancel_workflow.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		return err
+	}
+
+	err = workflow.ExecuteActivity(ctx, cancelActivity, "operationToCancel---cancel workflow after start. Using command 'cadence --do samples-domain wf cancel -w <WorkflowID>' ", 240).Get(ctx, nil)
+	if err != nil {
+		if cadence.IsCanceledError(err) {
+			// here, we need to get a new ctx to perform cleanup
+			newCtx, _ := workflow.NewDisconnectedContext(ctx)
+			// MyCancelActivity will delete that record from DB
+			err = workflow.ExecuteActivity(newCtx, cancelActivity, "cleanUp", 1).Get(ctx, nil)
```
Comment (mfateev) [inline cmd/samples/recipes/cancel/cancel_workflow.go]: It is super confusing to use the same activity for both logic to be cancelled and for the cleanup

File: cmd/samples/recipes/cancel/cancel_workflow.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+				return err
+			}
+		} else {
+			logger.Error("Activity failed.", zap.Error(err))
+			return err
+		}
+	}
+
+	//try to use old ctx, it won't be executed. Instead, it will return an canceled error
+	err = workflow.ExecuteActivity(ctx, cancelActivity, "tryOldContext", 1).Get(ctx, nil)
```
Comment (mfateev) [inline cmd/samples/recipes/cancel/cancel_workflow.go]: Do not use the same activity

File: cmd/samples/recipes/cancel/cancel_workflow.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func cancelActivity(ctx context.Context, name string, waitSecs int) error {
+	fmt.Println("first line in cancelActivity " + name)
+	logger := activity.GetLogger(ctx)
+	logger.Info("activity started")
+	select {
+	case <-time.After(time.Duration(waitSecs) * time.Second):
+		logger.Info("wake up after sleep for " + strconv.Itoa(waitSecs) + " seconds")
+	case <-ctx.Done():
```
Comment (mfateev) [inline cmd/samples/recipes/cancel/cancel_workflow.go]: This is not going to work. Only activities that heartbeat can be cancelled.

File: cmd/samples/recipes/cancel/cancel_workflow.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+			logger.Error("Activity failed.", zap.Error(err))
+			return err
+		}
+	}
+
+	//try to use old ctx, it won't be executed. Instead, it will return an canceled error
+	err = workflow.ExecuteActivity(ctx, cancelActivity, "tryOldContext", 1).Get(ctx, nil)
+	if err != nil {
+		logger.Error("Activity failed, this is expected", zap.Error(err))
+		// if returning this err here, the workflow will finished as Canceled.
```
Comment (mfateev) [inline cmd/samples/recipes/cancel/cancel_workflow.go]: We want workflow finished as canceled as we used the CLI to cancel it.

File: cmd/samples/recipes/cancelactivity/workflow.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	defer func() {
+		newCtx, _ := workflow.NewDisconnectedContext(ctx)
+		err := workflow.ExecuteActivity(newCtx, cleanupActivity).Get(ctx, nil)
+		if err != nil {
+			logger.Error("Cleanup activity failed", zap.Error(err))
+		}
+	}()
+
+	var result string
+	err := workflow.ExecuteActivity(ctx, activityToBeCacneled).Get(ctx, &result)
```
Comment (yux0) [inline cmd/samples/recipes/cancelactivity/workflow.go]: activityToBeCanceled

File: cmd/samples/recipes/cancelactivity/workflow.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	logger.Info(fmt.Sprintf("activityToBeCacneled returns %v, %v", result, err))
```
Comment (yux0) [inline cmd/samples/recipes/cancelactivity/workflow.go]: activityToBeCanceled

File: cmd/samples/recipes/cancelactivity/main.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	h.StartWorkflow(workflowOptions, Workflow)
+}
+
+func cancelWorkflow(h *common.SampleHelper, wid string) {
+	h.CancelWorkflow(wid)
+}
+
+func main() {
+	var mode, wid string
+	flag.StringVar(&mode, "m", "trigger", "Mode is worker or trigger.")
```
Comment (mfateev) [inline cmd/samples/recipes/cancelactivity/main.go]: Add "cancel"

File: cmd/samples/recipes/cancelactivity/workflow.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		StartToCloseTimeout:    time.Minute * 30,
+		HeartbeatTimeout:       time.Second * 5,
+		WaitForCancellation:    true,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+	logger := workflow.GetLogger(ctx)
+	logger.Info("cancel workflow started")
+
+	defer func() {
+		newCtx, _ := workflow.NewDisconnectedContext(ctx)
```
Comment (mfateev) [inline cmd/samples/recipes/cancelactivity/workflow.go]: Add comment why disconnected

File: cmd/samples/recipes/cancelactivity/workflow.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	_ = workflow.ExecuteActivity(ctx, activityToBeSkipped).Get(ctx, nil)
+
+	logger.Info("Workflow completed.")
+
+	return nil
+}
+
+func activityToBeCanceled(ctx context.Context) (string, error) {
+	logger := activity.GetLogger(ctx)
+	logger.Info("activity started, you can use ./cancelactivity -m cancel -w <WorkflowID> or CLI: 'cadence --do samples-domain wf cancel -w <WorkflowID>' to cancel")
```
Comment (mfateev) [inline cmd/samples/recipes/cancelactivity/workflow.go]: to cancel workflow

File: cmd/samples/recipes/cancelactivity/workflow.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+		if err != nil {
+			logger.Error("Cleanup activity failed", zap.Error(err))
+		}
+	}()
+
+	var result string
+	err := workflow.ExecuteActivity(ctx, activityToBeCanceled).Get(ctx, &result)
+	logger.Info(fmt.Sprintf("activityToBeCanceled returns %v, %v", result, err))
+
+	_ = workflow.ExecuteActivity(ctx, activityToBeSkipped).Get(ctx, nil)
```
Comment (mfateev) [inline cmd/samples/recipes/cancelactivity/workflow.go]: I wouldn't ignore error here and print its value as well

Comment (longquanzheng) [reply cmd/samples/recipes/cancelactivity/workflow.go]: done

File: cmd/samples/recipes/cancelactivity/workflow.go (hunk 9 above)
Comment (longquanzheng) [reply cmd/samples/recipes/cancelactivity/workflow.go]: done

File: cmd/samples/recipes/cancelactivity/workflow.go (hunk 8 above)
Comment (longquanzheng) [reply cmd/samples/recipes/cancelactivity/workflow.go]: done

File: cmd/samples/recipes/cancelactivity/main.go (hunk 7 above)
Comment (longquanzheng) [reply cmd/samples/recipes/cancelactivity/main.go]: done

**Pull request merged**